test/integration-test/src/tests/tcx.rs

```diff
@@ -0,0 +1,178 @@
+
+use crate::{
+    tests::log::{CapturedLog, TestingLogger},
+    utils::NetNsGuard,
+};
+
+fn setup_logs(loader: &mut Ebpf, logs: &Arc<Mutex<Vec<CapturedLog<'static>>>>) {
+    let captured_logs = logs.clone();
+    EbpfLogger::init_with_logger(
+        loader,
+        TestingLogger {
+            log: move |record: &Record| {
+                let mut logs = captured_logs.lock().unwrap();
+                logs.push(CapturedLog {
+                    body: format!("{}", record.args()).into(),
+                    level: record.level(),
+                    target: record.target().to_string().into(),
+                    timestamp: Some(Instant::now()),
+                });
+            },
+        },
+    )
+    .unwrap();
+}
+
+#[test(tokio::test)]
+async fn tcx_ordering() {
+    let kernel_version = KernelVersion::current().unwrap();
+    if kernel_version < KernelVersion::new(6, 6, 0) {
+        eprintln!("skipping tcx_ordering test on kernel {kernel_version:?}");
+        return;
+    }
+
+    let _netns = NetNsGuard::new();
+
+    let mut program0 = EbpfLoader::new()
+        .set_global("ORDER", &0, true)
+        .load(crate::TCX)
+        .unwrap();
+    let mut program1 = EbpfLoader::new()
+        .set_global("ORDER", &1, true)
+        .load(crate::TCX)
+        .unwrap();
+    let mut program2 = EbpfLoader::new()
+        .set_global("ORDER", &2, true)
+        .load(crate::TCX)
+        .unwrap();
+    let mut program3 = EbpfLoader::new()
+        .set_global("ORDER", &3, true)
+        .load(crate::TCX)
+        .unwrap();
+
+    let logs0: Arc<Mutex<Vec<CapturedLog>>> = Arc::new(Mutex::new(Vec::new()));
+    setup_logs(&mut program0, &logs0);
+
+    let logs1 = Arc::new(Mutex::new(Vec::new()));
+    setup_logs(&mut program1, &logs1);
+
+    let logs2 = Arc::new(Mutex::new(Vec::new()));
+    setup_logs(&mut program2, &logs2);
+
+    let logs3 = Arc::new(Mutex::new(Vec::new()));
+    setup_logs(&mut program3, &logs3);
+
+    let prog0: &mut SchedClassifier = program0
+        .program_mut("tcx_order")
+        .unwrap()
+        .try_into()
+        .unwrap();
+    prog0.load().unwrap();
+
+    let prog1: &mut SchedClassifier = program1
+        .program_mut("tcx_order")
+        .unwrap()
+        .try_into()
+        .unwrap();
+    prog1.load().unwrap();
+
+    let prog2: &mut SchedClassifier = program2
+        .program_mut("tcx_order")
+        .unwrap()
+        .try_into()
+        .unwrap();
+    prog2.load().unwrap();
+
+    let prog3: &mut SchedClassifier = program3
+        .program_mut("tcx_order")
+        .unwrap()
+        .try_into()
+        .unwrap();
+    prog3.load().unwrap();
+
+    // Test LinkOrder::first() and LinkOrder::set_expected_revision()
+    let mut order: LinkOrder = LinkOrder::last();
+    order.set_expected_revision(u64::MAX);
+    let options = TcAttachOptions::TcxOrder(order);
+    let result = prog0.attach_with_options("lo", TcAttachType::Ingress, options);
+    assert!(result.is_err());
```

Comment:
Why does this fail?

Reply:
And you're not testing that set_expected_revision(wrong_revision) fails after
some successful attachments

Reply:
I added two successful uses of expected_revision, followed by an unsuccessful use.